File: 一知半解/Hard/239. Sliding Window Maximum.py
class Solution:
    # 解題思緒，理論上會變動的只有最左跟最右，由於窗口會移動，有可能最大值傳遞的時候本體已經跌落窗口，因此要同時檢查最大值為多少以及是否該最大值還在window裡
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        # 如果長度1回傳本體
        if len(nums)==1:
            return nums
        # 做一個雙向的queue 來進行同時儲存最大值以及檢查位置，該queue負責儲存index
        ans=[];start = 0 ;end = 0;queue=[]
        for index,num in enumerate(nums):
            # 當queue內部還有東西，並且nums[queue[-1]]比現在的值小(代表queue的queue轉換成num 的最小值比現在的值還小)
            while queue and nums[queue[-1]]<=num:
                # 由於我們是要找最大值，因此我們要一直把最後一個踢出去，只要踢出去迴圈就會重新檢查一次，如果是小就在踢這樣就可以找出所有比這個小
                queue.pop()
            # 把現在的index往後加進去，這樣就符合前面一個數值一定比他大，或是上面清空後他是第一個值(最大)
            queue.append(index)
            # 如果長度不到window，跳過
            if index-start+1!=k:
                continue 
            # 把最大值加進去 
            ans.append(nums[queue[0]])
            # 如果發現最大值的index跟start同樣，代表他馬上就要消失了，所以把它扣掉
            if start==queue[0]:
                queue.pop(0)
            start+=1
        return ans


    # def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
    #     if len(nums)==1:
    #         return nums
    #     ans=[];start = 0 ;end = 0;queue=[]
    #     for index,num in enumerate(nums):
    #         while queue and queue[-1]<=num:
    #             queue.pop()
    #         queue.append(index)
    #         if index-start+1!=k:
    #             continue 
    #         ans.append(nums[queue[0]])
    #         if nums[start]==queue[0]:
    #             queue.pop(0)
    #         start+=1
    #     return ans
             

File: C++/一知半解/Hard/239. Sliding Window Maximum.cpp
#include<iostream>
#include <bits/stdc++.h>
#include <map>
#include <deque>
using namespace std;

//這題思路很簡單，但我寫道LTE，所以要更了解montonic stack的觀念
//要記得，我們維護montonic stack除了確保最裡面是最大值外，還要把之前比他小的每一個都刪掉
//這同時表示什麼，因為montonic的設定，比最大值晚進來的值會不管大小在他後面，因此我們pop左邊的值根本不需要搜索整個Montonic 
//我們只需要檢查最大值如果符合，代表這個是在左邊沒被擠掉的，如果不是，那他早就被擠掉了沒差。你不用去找他

//示意圖，[3,4,5,1,2,6,9,10,11,7,12],deque<int> mq;
//首先，把3加進去，好mq={3},再把4加進去(從前面加)，此時發現3比4小，把3刪掉，mq={4}，接下來一樣，5進去，mq={5}
//然後插入1，1比5小，擠不出去，因此mq={1,5},再來2，他比1大但比5小，把1擠掉,mq={2,5},6跟9跟10跟11就不一樣了
//6比2、5大，此刻2、5會被擠出去。mq={6},然後到11前都一樣， mq={11},然後7比11小，mq={7,11},最後12加入，mq={12}
//有發現嗎，單調維護的montonic stack，從前面加入，會不斷保證最後一個是最大值，此時如果我長度設定為2字串在1、2的時候
//mq長什麼樣子，他長成{1,5}，此時要移除的5會在最後面，其他的都會被前面擠出去用這個思路即可


class MontonicQueues{
    private:
        std::deque<int> montonicQueues;
    public:
        //最大值等於最後一個
        int max(){
            return montonicQueues.back();
        }
        //推進去才要處理，如果是空的直接推進去，跳出
        void push(int num){
            if (montonicQueues.empty()){
                montonicQueues.emplace_front(num);
                return ;
            }
            //不是空的且目前最前面的數字還比當前數字小，代表可以把他擠掉
            while(!montonicQueues.empty() && montonicQueues.front()<num){
                //把他擠掉
                montonicQueues.pop_front();
            }
            //擠完後把他加入正確的位置
            montonicQueues.emplace_front(num);
        }
        //照剛剛說的，要被移除的只有最大值還在的地方，更新的時候在檢查，是最大值的話把它移除
        void pop(int num){
            if(num == montonicQueues.back()){
                montonicQueues.pop_back();
            }
        }
};
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        int left=0;int right =0;std::vector<int> answer;MontonicQueues montonicQueue;
        if (nums.size()==k){
            int maxNum = *max_element(nums.begin(),nums.end());
            answer.emplace_back(maxNum);
            return answer;
        }
        for(int index=0;index<k;index++){
            montonicQueue.push(nums[index]);
        }

        answer.emplace_back(montonicQueue.max());
        for(right=k;right<nums.size();right++){
            montonicQueue.pop(nums[left]);
            montonicQueue.push(nums[right]);
            answer.emplace_back(montonicQueue.max());
            ++left;
        }
        return answer;        
    }
};

class Solution {
public:
    void push(int num,deque<int>& montonicQueues){
        vector<int> tempVec;
        if(montonicQueues.empty() || montonicQueues.back()<=num){
            montonicQueues.emplace_back(num);
            return;
        }
        int mright = montonicQueues.size()-1;int mleft = 0;int target = 0;
        while(mright>mleft){
            int mid = mleft+(mright-mleft)/2;
            if(montonicQueues[mid]>num){
                mright = mid;
            }
            else{
                mleft = ++mid;
            }
            target = mleft;
        }
        montonicQueues.emplace(montonicQueues.begin()+target,num);
        return;
    }
    void pop(int num,deque<int>& montonicQueues){
        vector<int> tempVec;
        if(montonicQueues.empty()){
            return;
        }
        if(montonicQueues.back()==num){
            montonicQueues.pop_back();
            return;
        }
        if(montonicQueues.front()==num){
            montonicQueues.pop_front();
            return;
        }
        int mright = montonicQueues.size()-1;int mleft = 0;int target = 0;
        while(mright>=mleft){
            int mid = mleft+(mright-mleft)/2;
            if(montonicQueues[mid]>num){
                mright = --mid;
            }
            else if(montonicQueues[mid]<num){
                mleft = ++mid;
            }
            else if(montonicQueues[mid]==num){
                target = mid;
                break;
            }
        }
        montonicQueues.erase(montonicQueues.begin()+target);
        return;
    }
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        int left=0;int right =0;std::deque<int> montonicQueues;std::vector<int> answer;
        if (nums.size()==k){
            int maxNum = *max_element(nums.begin(),nums.end());
            answer.emplace_back(maxNum);
            return answer;
        }
        for(int index=0;index<k;index++){
            push(nums[index],montonicQueues);
        }

        answer.emplace_back(montonicQueues.back());
        for(right=k;right<nums.size();right++){
            // cout<<montonicQueues.front()<<endl;
            pop(nums[left],montonicQueues);
            push(nums[right],montonicQueues);
            answer.emplace_back(montonicQueues.back());
            ++left;
        }
        return answer;        
    }
};